add python annotation files

ImpleLee opened the pull request:
These files are hand-written, and should be manually maintained in the future. Be careful about potential errors!

I tried several automatic `pyi` generation tools available on GitHub, but none of them is usable enough, so in the end I wrote them by hand (mostly by copy-and-paste).

`setup.py` is not changed in this file, because these files are automatically included (see [this documentation](https://setuptools.pypa.io/en/latest/userguide/miscellaneous.html)). However, this feature is experimental and may change in the future, so take your own risk. 

closes #532 .

ImpleLee committed:
Add Python type annotation files

ImpleLee committed:
Several improvements

add `-> None` for functions without return type;
remove `__new__` functions

ImpleLee committed:
Fix several issues

add some missing imports;
replace some C++ types with Python types;
reorganize some literals.

Adda0 commented on bindings/python/libmata/nfa/py.typed:
What is the purpose of this file?

koniksedy replied:
I believe it's there to show that the type hints are present.

ImpleLee replied:
I misunderstood the meaning of `namespace package` in the Python typing documentation.
After removing `nfa/py.typed`, the `pyi` files in `nfa` can still be distributed.

ImpleLee committed:
remove unnecessary `nfa/py.typed` file

This file was initially created because I misunderstood the meaning
of `namespace package` in the Python typing documentation (they need
separate `py.typed` files for each subpackage).

Adda0 reviewed (approved):
Looks good to me. Thank you for the contribution. This is a great addition.
